Framework/CurveFitting/src/Functions/CrystalFieldFunction.cpp

```diff
@@ -0,0 +1,1320 @@
+size_t
+CrystalFieldFunction::getParameterIndex(const ParameterReference &ref) const {
+  checkSourceFunction();
+  checkTargetFunction();
+  if (ref.getLocalFunction() == this) {
+    return ref.getLocalIndex();
+  }
+  auto index = m_control.getParameterIndex(ref);
+  if (index < m_nControlParams) {
+    return index;
+  }
+  index = m_source->getParameterIndex(ref);
+  if (index < m_source->nParams()) {
+    return index + m_nControlParams;
+  }
+  return m_target->getParameterIndex(ref) + m_nControlSourceParams;
+}
+
+/// Set up the function for a fit.
+void CrystalFieldFunction::setUpForFit() {
+  checkSourceFunction();
+  updateTargetFunction();
+  IFunction::setUpForFit();
+}
+
+/// Declare a new parameter
+void CrystalFieldFunction::declareParameter(const std::string &, double,
+                                            const std::string &) {
+  throw Kernel::Exception::NotImplementedError(
+      "CrystalFieldFunction cannot not have its own parameters.");
+}
+
+/// Build and cache the attribute names
+void CrystalFieldFunction::buildAttributeNames() const {
+  checkSourceFunction();
+  checkTargetFunction();
+  if (!m_attributeNames.empty()) {
+    return;
+  }
+  m_attributeNames = IFunction::getAttributeNames();
+  auto controlAttributeNames = m_control.getAttributeNames();
+  // Lambda function that moves a attribute name from controlAttributeNames
+  // to attNames.
+  auto moveAttributeName = [&](const std::string &name) {
+    auto iterFound = std::find(controlAttributeNames.begin(),
+                               controlAttributeNames.end(), name);
+    if (iterFound != controlAttributeNames.end()) {
+      controlAttributeNames.erase(iterFound);
+      m_attributeNames.push_back(name);
+    }
+  };
+  // Prepend a prefix to attribute names, ignore NumDeriv attribute.
+  auto prependPrefix =
+      [&](const std::string &prefix, const std::vector<std::string> &names) {
+        for (auto name : names) {
+          if (name == "NumDeriv")
+            continue;
+          name.insert(name.begin(), prefix.begin(), prefix.end());
+          m_attributeNames.push_back(name);
+        }
+      };
+  // These names must appear first and in this order in the output vector
+  moveAttributeName("Ions");
+  moveAttributeName("Symmetries");
+  moveAttributeName("Temperatures");
+  moveAttributeName("Background");
+  // Copy the rest of the names
+  m_attributeNames.insert(m_attributeNames.end(), controlAttributeNames.begin(),
+                          controlAttributeNames.end());
+  // Get
+  for (size_t iSpec = 0; iSpec < m_control.nFunctions(); ++iSpec) {
+    std::string prefix(SPECTRUM_PREFIX);
+    prefix.append(std::to_string(iSpec)).append(".");
+    auto names = m_control.getFunction(iSpec)->getAttributeNames();
+    for (auto &name : names) {
+      name.insert(name.begin(), prefix.begin(), prefix.end());
+    }
+    m_attributeNames.insert(m_attributeNames.end(), names.begin(), names.end());
+  }
+  auto nSpec = nSpectra();
```

Comment:
this doesn't seem to be used anywhere other than the for loop below. Could we not just put 
`for (size_t iSpec = nSpectra(); iSpec < m_target->nFunctions(); ++iSpec) {`